shin-core/src/format/scenario/info.rs

```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
     }
 }
 
+/// Matches a voice file to the lipsync character IDs for the characters speaking in the voice file, for lipsync purposes.
 #[derive(Debug, BinRead, BinWrite)]
 pub struct VoiceMappingInfoItem {
-    pub name_prefix: U16String,
-    pub unk1: U8List<u8>, // character id list for lipsync?
+    /// A pattern of voice file paths to be matched to the list of character IDs; either an individual path or a wildcard pattern specified using `*`. Does not include the `voice/` prefix or the file extension.
+    pub name_pattern: U16String,
+
+    /// List of character IDs for which a bustup with a matching lipsync character ID in its [`BustupInfoItem`] should have its lips animated if it is currently being displayed while a voice file matching the pattern is being played back.
+    pub lipsync_character_ids: U8List<u8>,
 }
 pub type VoiceMappingInfo = Vec<VoiceMappingInfoItem>;
 
+/// An entry in the Picture Box (`cgmode`).
+#[derive(Debug, BinRead, BinWrite)]
+pub struct PictureBoxInfoItem {
+    /// Internal name of the entry; defines the name of the texture to be loaded from `cgmode.txa` as the thumbnail for this entry.
+    pub name: U16String,
+
+    /// List of picture IDs that will be shown in sequence as the player clicks through the entry.
+    pub picture_ids: U16List<u16>,
+}
+pub type PictureBoxInfo = Vec<PictureBoxInfoItem>;
+
+/// An entry in the Music Box (`bgmmode`).
+#[derive(Debug, BinRead, BinWrite)]
+pub struct MusicBoxInfoItem {
+    /// The ID of the BGM track to be played if this entry is selected.
+    pub bgm_id: u16,
+
+    /// The index of the name to be displayed on the button for this entry, to be loaded from the `title*` textures in `bgmmode.txa`.
+    pub name_index: u16,
+
+    /// If this flag is set to 1, the BGM track will play only once instead of looping at the end. This is used e.g. for opening themes.
+    pub once_flag: u16,
+}
+pub type MusicBoxInfo = Vec<MusicBoxInfoItem>;
+
+trait FinalSegment {
```

Comment:
Also, maybe make a parsing-only helper do find the final segment? As far as I understand, all the `End` variants are signaled by `0` byte. Then you can have a universal type like this:

```rust
#[derive(BinRead, BinWrite)]
enum EndableSegment<T: BinRead + BinWrite> { // TODO: bounds are wrong, don't remember the correct ones
  #[brw(magic = 0x0u8)]
  End,
  Some(T),
}
```

This can be used in `parse_terminated_segment_list` and then discarded, producing just a `Vec` of segments

Reply:
Yes that's definitely a more elegant solution; changed it. (As a consequence the `End` segment no longer shows up in the parsed data, but I don't think that's really an issue)